libs/vkd3d/command.c

```diff
@@ -2692,6 +2696,339 @@ static void d3d12_command_list_handle_active_queries(struct d3d12_command_list *
+    {
+        const struct resolve_entry *r = &resolves[i];
+
+        VK_CALL(vkCmdCopyQueryPoolResults(list->vk_command_buffer,
+            r->query_pool, r->first_query, r->query_count,
+            resolve_buffer.buffer, resolve_buffer.offset + r->offset,
+            r->stride, VK_QUERY_RESULT_WAIT_BIT | VK_QUERY_RESULT_64_BIT));
+    }
+
+    /* Allocate scratch buffer for query lists */
+    entry_buffer_size = sizeof(struct query_entry) * list->pending_queries_count;
+
+    if (!d3d12_command_allocator_allocate_scratch_memory(list->allocator,
+            entry_buffer_size, ssbo_alignment, &entry_buffer))
+        goto cleanup;
+
+    if (!(query_map = vkd3d_malloc(sizeof(*query_map) * query_map_size)) ||
+            !(query_list = vkd3d_malloc(sizeof(*query_list) * list->pending_queries_count)))
+    {
+        ERR("Failed to allocate query map.\n");
+        goto cleanup;
+    }
+
+    /* Active list for the current dispatch */
+    src_queries = list->pending_queries;
+    dst_queries = query_list;
+
+    for (i = 0; i < dispatch_count; i++)
+    {
+        struct dispatch_entry *d = &dispatches[i];
+        memset(query_map, 0, sizeof(*query_map) * query_map_size);
+
+        /* First pass that counts unique queries since the compute
+         * shader expects list heads to be packed first in the array */
+        for (j = 0; j < d->virtual_query_count; j++)
+        {
+            const struct vkd3d_active_query *q = &src_queries[j];
+
+            if (!query_map[q->index])
+            {
+                query_map[q->index] = &dst_queries[d->unique_query_count++];
+                query_map[q->index]->dst_index = q->index;
+                query_map[q->index]->src_index = q->resolve_index;
+                query_map[q->index]->next = ~0u;
+            }
+        }
+
+        /* Second pass that actually generates the query list. */
+        for (j = 0, k = d->unique_query_count; j < d->virtual_query_count; j++)
+        {
+            const struct vkd3d_active_query *q = &src_queries[j];
+
+            /* Skip entries that we already added in the first pass */
+            if (query_map[q->index]->src_index == q->resolve_index)
+                continue;
+
+            query_map[q->index]->next = k;
+            query_map[q->index] = &dst_queries[k++];
+            query_map[q->index]->dst_index = q->index;
+            query_map[q->index]->src_index = q->resolve_index;
+            query_map[q->index]->next = ~0u;
+        }
+
+        src_queries += d->virtual_query_count;
+        dst_queries += d->virtual_query_count;
+    }
+
+    /* Upload query lists in chunks since vkCmdUpdateBuffer is limited to
+     * 64kiB per invocation. Normally, one single iteration should suffice. */
+    for (i = 0; i < list->pending_queries_count; i += 2048)
+    {
+        unsigned int count = min(2048, list->pending_queries_count - i);
+
+        VK_CALL(vkCmdUpdateBuffer(list->vk_command_buffer, entry_buffer.buffer,
```

Comment:
Isn't the scratch buffer host visible?

Reply:
No, scratch buffers are device local.

Reply:
Of course we can extend the scratch buffer system to allow host-visible buffers as well, but that's a larger change that would require a fair bit of work. Also we'd be doing random access on host memory in that case, which is probably bad.

Reply:
Right.